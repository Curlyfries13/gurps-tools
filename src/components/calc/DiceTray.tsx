import React, { useState } from 'react';
import { connect, ConnectedProps } from 'react-redux';

import { RootState } from 'src/redux';
import {
  setExpression,
  setValidExpression,
  setDiceMode,
} from 'src/redux/actions/damageActions';
import { basicPattern as dicePattern } from 'src/utils/diceUtils';
import { numberPattern } from 'src/utils/jsUtils';

// Consider if we want to move this out into Redux

const DiceTray = ({
  expression,
  expressionIsValid,
  setValidExpression,
  setExpression,
  setDiceMode,
}: Props) => {
  const id = 'dice-tray';
  const [valid, setValid]: [boolean, Function] = useState(false);

  function handleValueChange(event: React.ChangeEvent<HTMLInputElement>): void {
    // first try to parse as an integer
    const { value } = event.target;
    let outVal: number = parseInt(value, 10);
    setExpression(value);
    if (value.match(numberPattern) !== null && outVal >= 0) {
      setDiceMode(false);
      setValidExpression(true);
    } else if (value.match(dicePattern) !== null) {
      setDiceMode(true);
      setValidExpression(true);
    } else {
      // TODO: check if the expression is too long
      setDiceMode(false);
      setValidExpression(false);
    }
  }

  return (
    <>
      <label
        className='col col-form-label d-flex justify-content-center'
        htmlFor={id + 'value'}
      >
        Damage Value
      </label>
      <input
        type='tet'
        aria-label='value for damage calculation'
        className='form-control'
        id={id + 'value'}
        value={expression}
        onChange={handleValueChange}
      />
      <div className='row m-2'></div>
    </>
  );
};

function mapStateToProps(state: RootState) {
  return {
    expression: state.damageExpression,
    expressionIsValid: state.expressionIsValid,
  };
}

const mapDispatchToProps = {
  setValidExpression,
  setExpression,
  setDiceMode,
};

const connector = connect(mapStateToProps, mapDispatchToProps);
type Props = ConnectedProps<typeof connector>;

export default connector(DiceTray);
