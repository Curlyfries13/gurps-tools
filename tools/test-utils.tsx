// Using code from https://redux.js.org/recipes/writing-tests#connected-components
import React from 'react';

// test-utils.js
import { render } from '@testing-library/react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
// Import your own reducer
//
// console.log(React.createElement);

import * as reducer from 'src/redux/reducers';


const customRender = (
  ui,
  {
    initialState,
    store = createStore(reducer, initialState),
    ...renderOptions
  } = {}
) => {
  const Wrapper =({ children }) => {
    return <Provider store={store}>{children}</Provider>;
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
