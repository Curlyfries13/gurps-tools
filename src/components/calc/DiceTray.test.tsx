import React from 'react';

import { RenderResult } from '@testing-library/react';
import DiceTray from './DiceTray';

import { render, fireEvent, waitFor, screen } from 'tools/test-utils';
import store from 'src/redux';

it('matches snapshot', async () => {
  const component: RenderResult = render(<DiceTray />);
  expect(component).toMatchSnapshot();
});
