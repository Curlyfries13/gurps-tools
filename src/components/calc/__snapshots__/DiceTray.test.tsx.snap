// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <label
        class="col col-form-label d-flex justify-content-center"
        for="dice-trayvalue"
      >
        Damage Value
      </label>
      <div
        class="row m-2"
      >
        <div
          class="col-md-6"
        >
          <input
            aria-label="value for damage calculation"
            class="form-control"
            id="dice-trayvalue"
            type="tet"
            value=""
          />
        </div>
        <div
          class="col-md-6"
        >
          <select
            aria-label="damage type selection"
            class="form-select"
          >
            <option
              value="cr"
            >
              cr
            </option>
            <option
              value="cut"
            >
              cut
            </option>
            <option
              value="imp"
            >
              imp
            </option>
            <option
              value="pi"
            >
              pi
            </option>
            <option
              value="pi+"
            >
              pi+
            </option>
            <option
              value="pi++"
            >
              pi++
            </option>
            <option
              value="pi-"
            >
              pi-
            </option>
            <option
              value="burn"
            >
              burn
            </option>
            <option
              value="cor"
            >
              cor
            </option>
            <option
              value="fat"
            >
              fat
            </option>
            <option
              value="tox"
            >
              tox
            </option>
          </select>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <label
      class="col col-form-label d-flex justify-content-center"
      for="dice-trayvalue"
    >
      Damage Value
    </label>
    <div
      class="row m-2"
    >
      <div
        class="col-md-6"
      >
        <input
          aria-label="value for damage calculation"
          class="form-control"
          id="dice-trayvalue"
          type="tet"
          value=""
        />
      </div>
      <div
        class="col-md-6"
      >
        <select
          aria-label="damage type selection"
          class="form-select"
        >
          <option
            value="cr"
          >
            cr
          </option>
          <option
            value="cut"
          >
            cut
          </option>
          <option
            value="imp"
          >
            imp
          </option>
          <option
            value="pi"
          >
            pi
          </option>
          <option
            value="pi+"
          >
            pi+
          </option>
          <option
            value="pi++"
          >
            pi++
          </option>
          <option
            value="pi-"
          >
            pi-
          </option>
          <option
            value="burn"
          >
            burn
          </option>
          <option
            value="cor"
          >
            cor
          </option>
          <option
            value="fat"
          >
            fat
          </option>
          <option
            value="tox"
          >
            tox
          </option>
        </select>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
